### PR TITLE
dialects: (stencil) add StencilType constraints

### PR DESCRIPTION
**Pull request by superlopuh**
This adds some constraint constants and a new `.constr` helper to construct parametrized constraints for stencil types. Pyright errors 143 -> 133

### CODE REVIEW

File: xdsl/transforms/csl_stencil_to_csl_wrapper.py [hunk 1]
```diff
@@ -84,15 +84,16 @@ def match_and_rewrite(self, op: func.FuncOp, rewriter: PatternRewriter, /):
                     apply_op.done_exchange.block.args[1].type.get_shape()[-1],
                 )
 
-            # retrieve z_dim from done_exchange arg[0]
-            if isa(
+            # retrieve z_dim from post_process arg[0]
```
Comment (superlopuh) [inline xdsl/transforms/csl_stencil_to_csl_wrapper.py:87]:
```suggestion
            # retrieve z_dim from done_exchange arg[0]
```

File: xdsl/transforms/csl_stencil_to_csl_wrapper.py [hunk 2]
```diff
@@ -85,14 +85,15 @@ def match_and_rewrite(self, op: func.FuncOp, rewriter: PatternRewriter, /):
...
-            if isa(
+            if isattr(
                 field_t := apply_op.done_exchange.block.args[0].type,
-                stencil.StencilType[
-                    TensorType[Attribute] | memref.MemRefType[Attribute]
-                ],
+                stencil.StencilTypeConstr,
+            ) and isattr(
+                el_type := field_t.element_type,
+                AnyTensorTypeConstr | AnyMemRefTypeConstr,
```
Comment (superlopuh) [inline xdsl/transforms/csl_stencil_to_csl_wrapper.py:93]: this is the only kind of functional change, splitting the check into two, it should still be equivalent

**Review by math-fehr — APPROVED** [review]
Nice!
I think we should try to follow this on the other dialects later on

File: xdsl/transforms/csl_stencil_to_csl_wrapper.py (hunk 2 above)
Comment (n-io) [reply xdsl/transforms/csl_stencil_to_csl_wrapper.py:93]: LGTM, thanks for doing this

**Review by n-io — APPROVED** [review]

**Pull request merged**